### PR TITLE
Makes GetCurrendDir into a ReadFS effect

### PR DESCRIPTION
**Pull request by periodic**
# Overview

This makes an effect for `GetCurrentDir` as part of `ReadFS`.  This was one of the simplest changes identified in an inventory of our `sendIO` usage.

The main target here is config-file loading, which I standardized a bit so that each file wasn't reading in the directory independently.  For that I introduced a function to load the config file with the current dir as the default, which most commands use.

Other than that there was some usage for finding the directory to run Syft.

## Acceptance criteria

- There are fewer usages of `getCurrentDir` in the codebase.
- Other places we get the current dir are updated to use `ReadFS`

## Testing plan

One change here is that we now always print the config file as debug information for all the commands, instead of just a few.  This made it easy to run various comnmands and make sure that it was working in the new common code path.

I also added unit tests for `analyze` and `test` config loading.

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json`. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).

### CODE REVIEW

File: src/Control/Effect/Replay/TH.hs [hunk 1]
```diff
@@ -73,15 +73,24 @@ decodeSingle keyVal valVal con = do
...
-      noBindS [e|pure (EffectResult $(conAppliedE) $(varE resultNm))|]
-    ]
+    ( argBindingsE
+        <> [
+             -- decodedResultValue <- fromRecordedValue valVal
+             bindS (sigP (varP resultNm) (getGadtTy con)) [e|fromRecordedValue $(varE valVal)|]
+           , -- pure (EffectResult (Bar a b) decodedResultValue)
+             noBindS [e|pure (EffectResult $(conAppliedE) $(varE resultNm))|]
+           ]
+    )
```
Comment (scruffystuffs) [inline src/Control/Effect/Replay/TH.hs:93]: Is this change related?  I'd prefer to leave it out if not, TH code is really hard to review, I don't want to block the new effect constructor for this.

File: src/Effect/Exec.hs [hunk 2]
```diff
@@ -218,6 +220,15 @@ execThrow dir cmd = context ("Running command '" <> cmdName cmd <> "'") $ do
...
     Right stdout -> pure stdout
 
+-- | A variant of 'execThrow' that runs the command in the currend directory
+execThrow' :: (Has Exec sig m, Has ReadFS sig m, Has Diagnostics sig m) => Command -> m BL.ByteString
+execThrow' cmd = context ("Running command '" <> cmdName cmd <> "'") $ do
+  dir <- getCurrentDir
+  result <- exec dir cmd
+  case result of
+    Left failure -> fatal (CommandFailed failure)
+    Right stdout -> pure stdout
```
Comment (scruffystuffs) [inline src/Effect/Exec.hs:230]: DRY:
```suggestion
  execThrow dir cmd
```

File: src/Effect/ReadFS.hs [hunk 3]
```diff
@@ -351,6 +367,9 @@ runReadFSIO = interpret $ \case
   -- NB: these never throw
   DoesFileExist file -> sendIO (Directory.doesFileExist (fromSomeFile file))
   DoesDirExist dir -> sendIO (Directory.doesDirectoryExist (fromSomeDir dir))
+  GetCurrentDir -> do
+    PIO.getCurrentDir
+      `catchingIO` CurrentDirError
```
Comment (scruffystuffs) [inline src/Effect/ReadFS.hs:372]: If this throws on some occasions, put it above the NB comment.

File: src/App/Fossa/Config/LinkUserBinaries.hs [hunk 4]
```diff
@@ -94,9 +93,8 @@ loadConfig ::
   ) =>
   LinkUserBinsOpts ->
   m (Maybe ConfigFile)
-loadConfig LinkUserBinsOpts{..} = do
-  curdir <- sendIO getCurrentDir
-  resolveConfigFile curdir $ optConfig commons
+loadConfig =
+  resolveLocalConfigFile . optConfig . commons
```
Comment (scruffystuffs) [inline src/App/Fossa/Config/LinkUserBinaries.hs:97]:
```suggestion
loadConfig = resolveLocalConfigFile . optConfig . commons
```

File: src/App/Fossa/Config/Container.hs [hunk 5]
```diff
@@ -78,8 +77,7 @@ loadConfig = \case
   ContainerDumpScan _ -> pure Nothing
   -- Only parse config file if we're running analyze or test
   cmd -> do
-    curdir <- sendIO getCurrentDir
-    resolveConfigFile curdir $ getCfgFilePath cmd
+    resolveLocalConfigFile $ getCfgFilePath cmd
```
Comment (scruffystuffs) [inline src/App/Fossa/Config/Container.hs:80]: Remove `do` and one-line this

**Review by scruffystuffs — CHANGES_REQUESTED** [review]
Mostly nits, but I'd like to undo the TH changes if possible.  If not, can you leave a comment in the PR explaining why we had to make a change there?

File: src/Control/Effect/Replay/TH.hs (hunk 1 above)
Comment (periodic) [reply src/Control/Effect/Replay/TH.hs:93]: Oh, I forgot to explain that.  The TH we have doesn't seem to support no-arg constructors.  It ended up with missing the ReplayableData instance on `Unit(String)`.  I'm not sure what that means though, and it didn't seem necessary to match anyway so I did this instead.

I'll take another look tomorrow.

**Comment by periodic** [comment]
Did all the little changes.  Still have to figure out if the TH can be removed.

File: src/Control/Effect/Replay/TH.hs (hunk 1 above)
Comment (periodic) [reply src/Control/Effect/Replay/TH.hs]: I removed the `deriveReplayable` and reverted this, as you suggested.  I don't think it was correct anyway.

File: src/App/Fossa/Config/ConfigFile.hs [hunk 6]
```diff
@@ -87,6 +101,7 @@ resolveConfigFile ::
   Maybe FilePath ->
   m (Maybe ConfigFile)
 resolveConfigFile base path = do
+  logDebug $ "Loading configuration file from " <> pretty (show base)
```
Comment (scruffystuffs) [inline src/App/Fossa/Config/ConfigFile.hs:104]: [nit]: use helper
```suggestion
  logDebug $ "Loading configuration file from " <> viaShow base
```

File: src/Effect/Exec.hs [hunk 7]
```diff
@@ -39,9 +40,10 @@ import Data.String.Conversion (decodeUtf8, toString, toText)
 import Data.Text (Text)
 import Data.Text qualified as Text
 import Data.Void (Void)
+import Effect.ReadFS (ReadFS, getCurrentDir)
 import GHC.Generics (Generic)
 import Path
-import Path.IO
+import Path.IO (AnyPath (makeAbsolute))
```
Comment (scruffystuffs) [inline src/Effect/Exec.hs:46]: Let's make all imports explicit here

File: test/App/Fossa/Config/Utils.hs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+-- | Tests that the config loader uses the directory set in the arguments
+itShouldLoadFromTheConfiguredBaseDir ::
+  Parser opts -> (opts -> EffectStack (Maybe ConfigFile)) -> Spec
+itShouldLoadFromTheConfiguredBaseDir parser loadConfig =
+  it' "should load from the configured base dir"
+    . withTempDir "AnalyzeSpec"
+    $ \tempDir ->
+      do
+        sendIO $ BS.writeFile (toFilePath (tempDir </> $(mkRelFile ".fossa.yml"))) "version: 42\n"
```
Comment (scruffystuffs) [inline test/App/Fossa/Config/Utils.hs:29]: [suggestion]: if you use `version: 3`, you'll be able to use `mempty` for the expected config file value.

**Review by scruffystuffs — APPROVED** [review]
Few nits, but this looks good.  another effect in the bag.

File: src/Effect/Exec.hs (hunk 7 above)
Comment (periodic) [reply src/Effect/Exec.hs:46]: You sure? It's a lot. :D

File: test/App/Fossa/Config/Utils.hs (hunk 8 above)
Comment (periodic) [reply test/App/Fossa/Config/Utils.hs:29]: I want to be sure that we don't also default the version because we can't find the config.  I want to be sure that the version came from the file I created so I set it to an unusual value.

Comment (scruffystuffs) [reply test/App/Fossa/Config/Utils.hs:29]: We don't have a default config file, we handle a `Maybe ConfigFile` everywhere

Comment (scruffystuffs) [reply test/App/Fossa/Config/Utils.hs:29]: This is not a big deal, just making clear that this suggestion would be semantically safe.

Comment (periodic) [reply test/App/Fossa/Config/Utils.hs:29]: That's true.  I think this makes it abundantly clear that there's no way it's falling back to defaults so it's less likely the test could give a false positive.

Comment (periodic) [reply test/App/Fossa/Config/Utils.hs:29]: In a way what I want to test is that the config file is read, not that the config file matches a default, so I think this better describes the test.

We should have some tests about the default value though. :D

Comment (scruffystuffs) [reply test/App/Fossa/Config/Utils.hs:29]: If you just want to test that the file is read, you should use `isJust` instead.  That way the value doesn't matter at all.  As I said, we NEVER fallback to a default config file.
